addons/resource/models/resource.py

```diff
@@ -317,69 +278,61 @@ def get_working_intervals_of_day(self, cr, uid, id, start_dt=None, end_dt=None,
-        if id is None:
-            working_interval = []
+        if not self:
             if default_interval:
                 working_interval = (start_dt.replace(hour=default_interval[0], minute=0, second=0),
                                     start_dt.replace(hour=default_interval[1], minute=0, second=0))
-            intervals = self.interval_remove_leaves(cr, uid, working_interval, work_limits, context=context)
+            intervals = self.interval_remove_leaves(working_interval, work_limits)
             return intervals
 
         working_intervals = []
-        tz_info = fields.datetime.context_timestamp(cr, uid, work_dt, context=context).tzinfo
-        for calendar_working_day in self.get_attendances_for_weekday(cr, uid, id, start_dt, context=context):
-            if context and context.get('no_round_hours'):
+        for calendar_working_day in self.get_attendances_for_weekday(start_dt):
+            if self.env.context.get('no_round_hours'):
                 min_from = int((calendar_working_day.hour_from - int(calendar_working_day.hour_from)) * 60)
                 min_to = int((calendar_working_day.hour_to - int(calendar_working_day.hour_to)) * 60)
-                dt_f = work_dt.replace(hour=int(calendar_working_day.hour_from), minute=min_from)
-                dt_t = work_dt.replace(hour=int(calendar_working_day.hour_to), minute=min_to)
+                working_interval = (
+                    work_dt.replace(hour=int(calendar_working_day.hour_from), minute=min_from),
+                    work_dt.replace(hour=int(calendar_working_day.hour_to), minute=min_to),
+                    calendar_working_day.id,
+                )
             else:
-                dt_f = work_dt.replace(hour=int(calendar_working_day.hour_from))
-                dt_t = work_dt.replace(hour=int(calendar_working_day.hour_to))
-
-            # adapt tz
-            working_interval = (
-                dt_f.replace(tzinfo=tz_info).astimezone(pytz.UTC).replace(tzinfo=None),
-                dt_t.replace(tzinfo=tz_info).astimezone(pytz.UTC).replace(tzinfo=None),
-                calendar_working_day.id
-            )
-            working_intervals += self.interval_remove_leaves(cr, uid, working_interval, work_limits, context=context)
-
+                working_interval = (
+                    work_dt.replace(hour=int(calendar_working_day.hour_from)),
+                    work_dt.replace(hour=int(calendar_working_day.hour_to)),
+                    calendar_working_day.id,
+                )
+            working_intervals += self.interval_remove_leaves(working_interval, work_limits)
         # find leave intervals
         if leaves is None and compute_leaves:
-            leaves = self.get_leave_intervals(cr, uid, id, resource_id=resource_id, context=context)
-
+            leaves = self._model.get_leave_intervals(self.env.cr, self.env.uid, self.id, resource_id=resource_id, context=self.env.context)
```

Comment:
We can directly call self.get_leave_intervals(resource_id=resource_id), and [flatten](https://github.com/odoo-dev/odoo/blob/master-resource-new-api-bba/openerp/tools/misc.py#L245) the output (because, after installation of [stock_calendar](https://github.com/odoo-dev/odoo/blob/master-resource-new-api-bba/addons/stock_calendar/resource.py#L32), it will wrap the resultant list into another list. It is caused by automatic API bridging. Old API arguments are passed here to avoid that bridging, but can be removed by flattening the output).
Also add a short comment explaining the use of flatten here.